server/src/visitors/YmlBaseVisitor.ts

```diff
@@ -1,20 +1,242 @@
+            return;
+        }
+        // There is already a semicolon, or we are not in a block, or we don't want to add semicolons.
+        // Adding a semicolon isn't necessary.
+        if (
+            !!node.SEMICOLON() ||
+            !(node.parent instanceof ActionBlockContext) ||
+            !this.docFormatSettings.semicolonWhenPossible
+        ) {
+            this.visitChildren(node);
+            return;
+        }
+
+        // Case of a node without children. May happen when there are syntax errors.
+        if (node.childCount <= 0) {
+            return;
+        }
+        // Get the last child of this node − the one that ends the instruction.
+        // We do this instead of doing this inside a classic `visitXXX(XXXContext)`
+        // because here is the only place where we have enough context to add semicolon.
+        const lastChild = node.getChild(node.childCount - 1);
+
+        if (lastChild === node.instruction_assignment()) {
+            this.addSemiColon(node.instruction_assignment());
+        } else if (lastChild === node.instruction_multivaluedAssignment()) {
+            this.addSemiColon(node.instruction_multivaluedAssignment());
+        } else if (lastChild === node.instruction_return()) {
+            this.addSemiColon(node.instruction_return());
+        } else if (lastChild === node.instruction_chainedCall()) {
+            this.addSemiColon(node.instruction_chainedCall());
+        } else if (lastChild === node.instruction_try_catch()) {
+            this.addSemiColon(node.instruction_try_catch());
+        } else if (lastChild === node.instruction_timeCounter()) {
+            this.addSemiColon(node.instruction_timeCounter());
+        } else {
+            // do nothing → the instruction doesn't require a semicolon.
+        }
+        // Continue visiting nodes to apply other fixes.
+        this.visitChildren(node);
+    }
+
+    public visitInstruction_if(node: Instruction_ifContext): void {
+        if (this.isDocumentFormatImpossible()) {
+            return;
+        }
+        if (!node.actionBlockOrInstruction().actionBlock()) {
+            this.visitChildren(node);
+            return;
+        }
+
+        const ifSymbolEnd = node.IF().symbol.stopIndex;
+        const openParStart = node.OPEN_PAR().symbol.startIndex;
+        let sameLine = node.IF().symbol.line === node.OPEN_PAR().symbol.line;
+        if (this.docFormatSettings.spaceBetweenKeywordAndParenthesis) {
+            this.setOneSpaceInterval(ifSymbolEnd, openParStart, sameLine);
+        } else {
+            this.removeSpaceInterval(ifSymbolEnd, openParStart);
+        }
+
+        const closeParEnd = node.CLOSE_PAR().symbol.stopIndex;
+        const startIfBody = node.actionBlockOrInstruction()._start.startIndex;
+        sameLine = node.CLOSE_PAR().symbol.line === node.actionBlockOrInstruction()._start.line;
+        this.setOneSpaceInterval(closeParEnd, startIfBody, sameLine);
+        this.visitChildren(node);
+    }
```

Comment:
Omg I thought most of the space-management stuff was built in and you just had to name a bunch of tokens, but you actually had to write heaps of boilerplate-ish stuff. :scream:

Reply:
Yeah. This was kinda boring. But now that it is done, I feel that adding an option will be a piece of cake.